### ISSUE TITLE
Exception: Error: Unicode could not encode unicode string 

### ISSUE THREAD

**Issue by BlauFeuer** [evennia/evennia#998]
#### Brief summary of issue / Description of requested feature:
When a non-unicode session is to be sent a unicode string, /evennia/utils/utils.py", line 593, in to_str raises an unhandled exception("Error: Unicode could not encode unicode string to a bytestring.")  

This can happen when any unicode character (glyph) is untranslatable into the non-unicode target encoding for a session. The glyph may be one that could degrade gracefully into ASCII, such as an ellipsis ( `…` to `...`) or an apostrophe ( `’` to  `'` ) or quotation marks, em or en-dashes (typically expressed as one or more hypens or minus sign).  

In some cases, lettering with diacritic marks could be stripped of them when untranslatable otherwise, in a crude fashion such as [this routine](https://gist.github.com/riccardomurri/3c3ccec30f037be174d3).

Some unicode glyphs will have no ASCII or latin-1 encoded character resembling it, By default they should be removed from the unicode string before being sent to the non-unicode target by a method that can be overriden in a game-specific way.

#### Steps to reproduce the issue / Reasons for adding feature:

1.  Use a non-unicode client (latin-1 is typical) to login, perhaps requiring `encoding latin-1` or `@option encoding=latin-1` command to set the session encoding.
2. Observe text being sent either from a unicode-enabled session or in text on an in-game object. Dashes ‒  –  —  ―, apostrophe ’, ellipsis … ⋯ or any of the "smart" quotes ‘ ’  “ ”  will do. Or the name `Sigurður`
3. Causes traceback with unhandled exception error.

Graceful degrading or substitution of untranslatable characters into `?` or similar is preferable to traceback generation and not sending any other characters in the unicode string to the non-unicode session.

#### Error output / Expected result of feature
`say My name is Sigurður.` yields `Exception: Error: Unicode could not encode unicode string 'Sigurður says, "My name is Sigurður."'(latin-1) to a bytestring.`  /evennia/utils/utils.py", line 593, in to_str raises an unhandled exception.

Could instead send: 'Sigur?ur says, "My name is Sigur?ur.' or be handled by game-specific code.

#### Extra information, such as Evennia revision/repo/branch, operating system and ideas for how to solve / implement: Latest Evennia master branch any operating system.

Any more graceful way would be good. By default, stripping unencodable characters, but best to allow customized methods.

**Comment by Griatch** [comment]
This sounds like the exact use case of the ENCODING `@option`...but sure, a Traceback should be avoided.

**Comment by BlauFeuer** [comment]
The latin-1-using client can set a utf-8 encoding and _avoid receiving the traceback error_, but then receives mistranslated characters (gremlins) that might otherwise be able to be reencoded properly (umlauted letters and various high-byte characters in the latin-1 encoding).

As it is, setting latin-1 encoding when the client only handles latin-1 does get the traceback when unicode-specific characters are sent.

There's no avoiding the situation until everyone's using unicode or an encoding that has all the glyphs that unicode has, or unicode sessions are limiting the sent characters to a common subset.  How it's handled beyond filtering unreencodable characters seem to me to be a game-specific issue.

**Comment by BlauFeuer** [comment]
Is there a way I can just provide my own exception handler routine instead of always shutting down the whole output line due to one character that one user session can't decode?  This has been a problem for months and I don't really know what to do about it in the game code. I've tried first checking the session encoding within an inline function, but the function is parsed after the exception, so that's clearly not going to work.

**Comment by Griatch** [comment]
No, the error must be caught and dealt with in-place inside the data path. So if you have a solution, a PR is the way to go.

**Comment by BlauFeuer** [comment]
See link in "lettering with diacritic marks could be stripped of them when untranslatable otherwise, in a crude fashion such as this routine." in original post.

**Comment by Griatch** [comment]
@BlauFeuer 

Pushed a tentative fix, check it out.

**Comment by Griatch** [comment]
As far as I can tell, this bug has been fixed. Closing.

**Comment by BlauFeuer** [comment]
@Griatch Thanks! I pulled this and people are happy that this resolves a say or pose with unicode-unique characters being sent from being silenced and generating a traceback every time. It also allows characters, rooms or objects to have unicode-unique glyphs in the name or description.

Thus, this now allows outgoing unicode messages to non-unicode sessions even though they might be garbled.  The best way to clean up the messages seems to be to detect the recipients and make substitutions before calling the `msg` method.  I tried allowing substitutions by way of an inline function, but that seems to be too late in the message chain. What are your thoughts? Should I pre-process all message calls or wrap them in some way, or overwrite the `msg` method with something that makes substitutions (effectively making a new markup type)?
